ts/Extensions/DataGrouping.ts

```diff
@@ -508,6 +528,128 @@ const groupData = function (
+
+            i = groupedXData.length - 1;
+            while (i-- && i > 0) {
+                groupedXData[i] += shiftInterval;
+            }
+        }
+
+        // Change the first point position, but only when it is
+        // the first point in the data set not in the current zoom.
+        if (
+            firstAnchor &&
+            firstAnchor !== 'start' &&
+            series.xData[0] >= groupedXData[0]
+        ) {
+            const firstGroupstEnd = (series.groupMap[0].start as number) +
+                (series.groupMap[0].length as number - 1);
+
+            groupedXData[0] = ({
+                middle: groupedXData[0] + 0.5 * totalRange,
+                end: groupedXData[0] + totalRange,
+                firstPoint: series.xData[0],
+                lastPoint: series.xData[firstGroupstEnd]
+            } as Highcharts.anchorChoiceType)[firstAnchor];
+        }
+
+        // Change the last point position but only when it is
+        // the last point in the data set not in the current zoom.
+        if (
+            lastAnchor &&
+            lastAnchor !== 'start' &&
+            groupedXData[groupedDataLength] >= xMax - (totalRange as number)
+        ) {
+            const lastGroupStart = series.groupMap[series.groupMap.length - 1].start;
+
+            groupedXData[groupedDataLength] = ({
+                middle: groupedXData[groupedDataLength] + 0.5 * totalRange,
+                end: groupedXData[groupedDataLength] + totalRange,
+                firstPoint: lastGroupStart && series.xData[lastGroupStart],
+                lastPoint: series.xData[series.xData.length - 1]
+            } as Highcharts.anchorChoiceType)[lastAnchor];
+        }
+    }
+};
+
+const adjustExtremes = function (
+    xAxis: Axis,
+    series: Series,
+    groupedXData: Array<number>
+): any {
+    const options = series.options,
+        dataGroupingOptions = options.dataGrouping,
+        visible = series.visible;
+
+    // Make sure the X axis extends to show the first group (#2533)
+    // But only for visible series (#5493, #6393)
+    if (
+        defined(groupedXData[0]) &&
+        groupedXData[0] < (xAxis.min as any) &&
+        visible
+    ) {
+        if (
+            (
+                !defined(xAxis.options.min) &&
+                (xAxis.min as any) <= (xAxis.dataMin as any)
+            ) ||
+            xAxis.min === xAxis.dataMin
+        ) {
+            xAxis.min = Math.min(groupedXData[0], (xAxis.min as any));
+        }
+
+        xAxis.dataMin = Math.min(
+            groupedXData[0],
+            (xAxis.dataMin as any)
+        );
+    }
+
+    // When the last anchor set, change the extremes that
+    // the last point is visible (#12455).
+    if (
+        defined(groupedXData[groupedXData.length - 1]) &&
+        groupedXData[groupedXData.length - 1] > (xAxis.max as any) &&
+        visible &&
+        dataGroupingOptions &&
+        dataGroupingOptions.lastAnchor !== 'start'
```

Comment:
Just a question: Why `max` needs a check for `lastAnchor = start`?

Reply:
It's not necessary to check that here, I removed that ;)